atomix/cluster/src/main/java/io/atomix/raft/snapshot/impl/FileBasedSnapshotStore.java

```diff
@@ -0,0 +1,318 @@
+    return getLatestSnapshot().map(PersistedSnapshot::getIndex).orElse(0L);
+  }
+
+  @Override
+  public void delete() {
+    currentPersistedSnapshotRef.set(null);
+
+    try {
+      LOGGER.debug("DELETE FOLDER {}", snapshotsDirectory);
+      FileUtil.deleteFolder(snapshotsDirectory);
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    try {
+      LOGGER.debug("DELETE FOLDER {}", pendingDirectory);
+      FileUtil.deleteFolder(pendingDirectory);
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private void observeSnapshotSize(final PersistedSnapshot persistedSnapshot) {
+    try (final var contents = Files.newDirectoryStream(persistedSnapshot.getPath())) {
+      var totalSize = 0L;
+
+      for (final var path : contents) {
+        if (Files.isRegularFile(path)) {
+          final var size = Files.size(path);
+          snapshotMetrics.observeSnapshotFileSize(size);
+          totalSize += size;
+        }
+      }
+
+      snapshotMetrics.observeSnapshotSize(totalSize);
+    } catch (final IOException e) {
+      LOGGER.warn("Failed to observe size for snapshot {}", persistedSnapshot, e);
+    }
+  }
+
+  private void purgePendingSnapshots(final long cutoffIndex) {
+    LOGGER.debug(
+        "Search for orphaned snapshots below oldest valid snapshot with index {} in {}",
+        cutoffIndex,
+        pendingDirectory);
+
+    try (final var pendingSnapshots = Files.newDirectoryStream(pendingDirectory)) {
+      for (final var pendingSnapshot : pendingSnapshots) {
+        purgePendingSnapshot(cutoffIndex, pendingSnapshot);
+      }
+    } catch (final IOException e) {
+      LOGGER.warn(
+          "Failed to delete orphaned snapshots, could not list pending directory {}",
+          pendingDirectory,
+          e);
+    }
+  }
+
+  private void purgePendingSnapshot(final long cutoffIndex, final Path pendingSnapshot) {
+    final var optionalMetadata = FileBasedSnapshotMetadata.ofPath(pendingSnapshot);
+    if (optionalMetadata.isPresent() && optionalMetadata.get().getIndex() < cutoffIndex) {
+      try {
+        FileUtil.deleteFolder(pendingSnapshot);
+        LOGGER.debug("Deleted orphaned snapshot {}", pendingSnapshot);
+      } catch (final IOException e) {
+        LOGGER.warn(
+            "Failed to delete orphaned snapshot {}, risk using unnecessary disk space",
+            pendingSnapshot,
+            e);
+      }
+    }
+  }
+
+  public Path getPath() {
+    return snapshotsDirectory;
+  }
+
+  @Override
+  public void close() {
+    listeners.clear();
+  }
+
+  private boolean isCurrentSnapshotNewer(final FileBasedSnapshotMetadata metadata) {
+    final var persistedSnapshot = currentPersistedSnapshotRef.get();
+    return (persistedSnapshot != null && persistedSnapshot.getId().compareTo(metadata) >= 0);
+  }
+
+  PersistedSnapshot newSnapshot(final FileBasedSnapshotMetadata metadata, final Path directory) {
```

Comment:
Shouldn't we synchronize access to this method?. There is a potential race condition between comparing current snapshot and assigning the new one when concurrent replication is in progress. Alternative use `currentPersistedSnapshotRef.compareAndSet`.

Reply:
You mean because raft leader-follower repl- and our zeebe replication?

Reply:
Lets shortly think about it and discuss it/

This can only happen for the same snapshot, right? Because we just have one snapshot at a time and this only happens when raft replicate for a slow follower **AND** our snapshot replication replicates the snapshot. 

If slow follower replication start for an old it would fail if we replace it in the meantime with a different one.

Say we have two calls which calling concurrently this method.  I would expect that one of the moves will fail because the snapshot dir already exists. In all cases the result is the same that the snapshot references the right snapshot or do I miss something?